src/tiled/abstractobjecttool.cpp

```diff
@@ -276,6 +288,24 @@ void AbstractObjectTool::showContextMenu(MapObjectItem *clickedObjectItem,
     Utils::setThemeIcon(removeAction, "edit-delete");
     Utils::setThemeIcon(propertiesAction, "document-properties");
 
+    QList<MapObjectItem*> underlyingObjects = listOfObjectItemsAt(scenePos);
+    if (underlyingObjects.size() > 1) {
+        menu.addSeparator();
+
+        QMenu *selectUnderlyingMenu = menu.addMenu(tr("Select an underlying object"));
+
+        for (int levelNum = 0; levelNum < underlyingObjects.size(); ++levelNum) {
+            const QString& objectName = underlyingObjects[levelNum]->mapObject()->name();
+            QString actionName = (objectName.isEmpty() ? tr("Object at level %n", "", levelNum + 1) : objectName)
+                    + tr(levelNum ? "" : " (topmost)");
+            QAction *action = selectUnderlyingMenu->addAction(actionName);
+            if (levelNum == 0)
+                action->setEnabled(false);//just to set a starting point
+            else
+                action->setData(QVariant::fromValue(levelNum));
```

Comment:
If you put `Q_DECLARE_METATYPE(Tiled::Internal::MapObjectItem*)` at the bottom of `mapobjectitem.h`, then you can do `action->setData(QVariant::fromValue(underlyingObjects[levelNum])` here and `action->data().value<MapObjectItem*>()` below.